### ISSUE TITLE
Transactions don't propagate when low peer count / not 100% synced  

### ISSUE THREAD

**Issue by D3m0nKingx** [stratisproject/StratisBitcoinFullNode#957]
Discovered if the node is connected with a low peer count and majority of the peers aren't 100% synced the Transactions don't propagate

```bash 
Node:::ffff:100.19.40.254:50290,         connected (inbound),       agent StratisBitcoin:1.0.7,  height=5000
Node:::ffff:13.64.76.48:26178,           connected (outbound),      agent /Stratis:2.0.0.3/,  height=154000
Peer:::ffff:88.253.146.132:62256,        connected (inbound),       agent Breeze:1.0.5,   height=226432
Peer:::ffff:176.24.25.143:47818,         connected (inbound),       agent StratisBitcoin:1.0.7,  height=231432
Peer:::ffff:62.194.112.119:56404,        connected (inbound),       agent StratisBitcoin:1.0.7,  height=96172
Peer:::ffff:212.92.108.174:54167,        connected (inbound),       agent StratisBitcoin:1.0.7,  height=44173```

**Comment by D3m0nKingx** [comment]
Possibly due to the peers being in IBD?

**Comment by noescape00** [comment]
The propagation algorithm works like this: 

1) select half of the peers and send them the transaction
2) wait other half to propagate it back. if they did or we received a block with out tx- consider it propagated. 

So mist likely that the half that was selected for sending the tx to was in IBD. 

Overall this bug could be solved by not including IBD peers in this algorithm. 
And right now we don't have a way for knowing if peer is in IBD. But this will be fixed after part 2 of [this ](https://github.com/stratisproject/StratisBitcoinFullNode/pull/956)PR series is done.

**Comment by quantumagi** [comment]
Regarding the propagation algorithm. Would this work?:

1.  Send transaction to two random peers giving preference to high quality score peers that are not in IBD.

2. Repeat (with some interval) until half of the peers are propagating the transaction back.

**Comment by D3m0nKingx** [comment]
Based on that algo that the FN is using, I'm not sure the logic/reasoning/security/benefit to having only 50% of peers getting broadcasted of a transaction? I observed how the Stratis-qt (and others) do this, and they are broadcasting to 100% of connected peers. 

@fassadlr could you possibly elaborate on this?

**Comment by noescape00** [comment]
@D3m0nKingx without the propagation to the half of the peers we son't have a way to know if tx was propagated or not until the block with this tx is mined. 
So it's a trade off- if the tx wasn't propagated we at least know for sure that we can respend the coins used in that tx. 
Also it is planned to add this feature in the full node UI- the user will be able to see al txes and manually remove\resend transactions that were not propagated.

**Comment by D3m0nKingx** [comment]
@noescape00 agreeable, furthermore, if 100% of the connected peers are broadcasted to, it's more likely for a transaction to get propagated I believe than just using half of them. 

In addition, from a security point of view, if we're only broadcasting to half of the connections, then an attacker could effectively bring down the network by 50% persay, which then reduces the number of usable peers by 50% as well (albeit those peers aren't dirty, etc).

**Comment by D3m0nKingx** [comment]
At some point a Network Stress Test should be performed to give some more clear picture of this.

**Comment by noescape00** [comment]
>In addition, from a security point of view, if we're only broadcasting to half of the connections, then an attacker could effectively bring down the network by 50% persay, which then reduces the number of usable peers by 50% as well (albeit those peers aren't dirty, etc).

If 50% of the network goes down we will loose 4\8 max peers. So we will just connect to 4 more peers in a minute and the propagation chances are the same as before.

**Comment by D3m0nKingx** [comment]
> If 50% of the network goes down we will loose 4\8 max peers. So we will just connect to 4 more peers in a minute and the propagation chances are the same as before.

Assuming there's no issues with the "4 more peers" ... 
Although theoretical there'd be such situation to cripple the network to that degree, without actually seeing an event as this. 

Either way, putting any ratio for propagation seems more of a performance degradation imho.

**Comment by fassadlr** [comment]
@D3m0nKingx I dont think I can comment here? What algo are you referring to?

**Comment by noescape00** [comment]
>Assuming there's no issues with the "4 more peers" ...

Current peer selection algho would try connecting mostly to known before peers so we should gain those +4 just in a few minutes if this scenario happens.

**Comment by D3m0nKingx** [comment]
@fassadlr 

> The propagation algorithm works like this:
>
> select half of the peers and send them the transaction
wait other half to propagate it back. if they did or we received a block with out tx- consider it propagated.

As stated by @noescape00

**Comment by fassadlr** [comment]
Ahhhh I see! I havent really been involved in that selection algo, so cant really comment at this point :) I'll need to go and investigate first.

**Comment by noescape00** [comment]
>Either way, putting any ratio for propagation seems more of a performance degradation imho.

Agreed. But this trade off is needed for the full node UI and the breeze wallet. Removing this trade off will make us remove a  potentially useful feature from the wallets. 

Though if you still think that propagating to 100% makes more sense- start a separate issue discussing the propagation algho and request a change- let's discuss this separately since it involves changing some very important logic.

**Comment by noescape00** [comment]
@fassadlr to investigate this take a look at wallet manager : SendTransactionAsync and at light wallet and full node's broadcaster managers.

**Comment by noescape00** [comment]
Also @Aprogiena is familiar with the propagation algorithm so you'd probably want him to comment on this as well.

**Comment by D3m0nKingx** [comment]
I started an issue specific to for the Algo discussion, lets continue the discussion there #973

**Comment by Aprogiena** [comment]
@noescape00 unfortunately not, i'm familiar with block propagation, not transaction propagation

**Comment by noescape00** [comment]
@Aprogiena ok, then it was Dan who explained the current algho to me, I forgot (tried to check in slack but it removed all the conversations).

**Comment by bokobza** [comment]
The reason we send to half the peers is because in the case of the light wallet, where we don't have the mempool, we can't verify transactions.
When we send the transaction to half the peers and that we receive it back from another peer, we know that it's been validated and passed on.
For the full node wallet, this could probably be increased to 100%.

**Comment by dangershony** [comment]
Agreed for fullnode we probably don't need to wait for peer propagation.
However we did decide on another issue to use a propagation flag and allow users to delete (at some point) trx that have not been marked as propagated for a long while.

**Comment by bokobza** [comment]
We can close this can't we @dangershony @fassadlr ?
